crates/composition/src/ingest_subgraph/object.rs

```diff
@@ -0,0 +1,42 @@
+use super::schema_definitions::FederationDirectivesMatcher;
+use crate::{subgraphs::DefinitionId, Subgraphs};
+use async_graphql_parser::types as ast;
+use async_graphql_value::ConstValue;
+
+/// Returns whether the object is shareable.
+pub(super) fn ingest_directives(
+    definition_id: DefinitionId,
+    type_definition: &ast::TypeDefinition,
+    subgraphs: &mut Subgraphs,
+    federation_directives_matcher: &FederationDirectivesMatcher<'_>,
+) -> bool {
+    let mut is_shareable = false;
+
+    for directive in &type_definition.directives {
+        if federation_directives_matcher.is_shareable(&directive.node.name.node) {
+            is_shareable = true;
+            continue;
+        }
+
+        if federation_directives_matcher.is_key(&directive.node.name.node) {
+            let fields_arg = directive.node.get_argument("fields").map(|v| &v.node);
+            let Some(ConstValue::String(fields_arg)) = fields_arg else {
+                continue;
+            };
+            let Ok(selection_id) = subgraphs.selection_set_from_str(fields_arg) else {
+                continue; // TODO: error handling in subgraph ingestion?
+            };
+            let is_resolvable = directive
+                .node
+                .get_argument("resolvable")
+                .and_then(|v| match v.node {
+                    ConstValue::Boolean(b) => Some(b),
+                    _ => None,
+                })
+                .unwrap_or(true); // defaults to true
+            subgraphs.push_key(definition_id, selection_id, is_resolvable);
```

Comment:
opt: usually not a fan of having both an output and an output argument, but I guess that'll be reworked X times anyway.

Reply:
Yes that felt dirty as well as I was writing it. I was writing down the reasoning for this particular case in this comment box and it's not strong. I'll change this function to do everything with the output argument.

Reply:
Ah no, it's because we don't store `is_shareable` inside the object, but rather inside the fields, to have a single source of truth, so it isn't in `Subgraphs`. I'll think about a clean way to incorporate that.